### PR TITLE
Implementation of target selection.

### PR DESCRIPTION
**Pull request by azieba**
There is a target selection key defined in the configuration, but it has no binding.
This is a proposal for implementation for it. It uses target list from Sensors.cpp.
The friend or foe identification is supposed to rely on relations which seem not to be implemented at the moment, I decided that checking if the target has active AI command KILL on the player is a good indication of being hostile. This improvement also turns ship speed lines to red for foes.
The selection works by picking nearest hostile if no combat target is selected. Next key press should select next hostile target according to distance. If the furthest hostile is selected next key press will deselect combat target. So it is more like toggling. It works only within distance used in Sensors so 100km.
**EDIT**
The cycling of hostiles has been moved to separate key and implementation of selecting targets near center of the view was added.

### CODE REVIEW

**Comment by impaktor** [comment]
>  if the target has active AI command KILL on the player is a good indication of being hostile. This improvement also turns ship speed lines to red for foes.

Spontaneous thought: maybe they should also have acted on that command, like actually tried to use any weapon system on the target, lest the computer is mind reading?

**Comment by azieba** [comment]
> Spontaneous thought: maybe they should also have acted on that command, like actually tried to use any weapon system on the target, lest the computer is mind reading?

Honestly I haven't thought about it. I just wanted it to work like in other games and be most useful for me as a player.
But if you look for rationale: the game uses some logic to give kill command to some ships based on the circumstances, so let's say that the ship computer can also analyze the same circumstances and heuristically determine hostiles.

**Comment by Web-eWorks** [comment]
> Spontaneous thought: maybe they should also have acted on that command, like actually tried to use any weapon system on the target, lest the computer is mind reading?

At this stage, there is no supporting gameplay to require such a feature or make it actually useful unfortunately.

@azieba I will do a code review later, but in general I applaud the effort! You've been on fire recently with Pioneer contributions!

It seems from your description that this code cycles through hostile ships with each key press, regardless of where they are relative to the player's view.

As I would define it, the original intent for target selection was that pressing the key should select the target (whether combat or navigation, with a preference to combat targets over nav targets) closest to the forward vector of the ship or the center of your screen (e.g. when freelooking to the side). This is meant to be a quick and ergonomic alternative to selecting moving targets via the relatively small box you would otherwise have to click on with the mouse.

If that behavior is not what this PR currently implements, I'd recommend that the functionality you've added receive its own "Cycle Hostiles" input key; if you feel like going above and beyond, you can also implement the mentioned behavior for the target select key too! :smile:

**Comment by azieba** [comment]
> @azieba I will do a code review later, but in general I applaud the effort! You've been on fire recently with Pioneer contributions!

I think I rediscovered that programming can be fun and give satisfaction. 
 
> If that behavior is not what this PR currently implements, I'd recommend that the functionality you've added receive its own "Cycle Hostiles" input key; if you feel like going above and beyond, you can also implement the mentioned behavior for the target select key too! smile

Yes I can give it it's own key. I will also try to go "above and beyond". Should I switch the PR to draft?

**Comment by impaktor** [comment]
If there's a misunderstanding of what the target selection button should do, and that it shouldn't be so smart as the current code implementation, then this might be salvaged for future AI/Crew + turret -functionality?

**Comment by impaktor** [comment]
> EDIT
> Is there a way to add a message conveniently to all translation files, or one must edit all of them?

@azieba No, we never edit all translation files, just edit the en.json file. The `description` field is for the translator.
https://pioneerwiki.com/wiki/Translations

**Comment by Web-eWorks** [comment]
> If there's a misunderstanding of what the target selection button should do, and that it shouldn't be so smart as the current code implementation, then this might be salvaged for future AI/Crew + [turret](https://github.com/fluffyfreak/pioneer/tree/new-turrets) -functionality?

@impaktor the "select target" button is generally understood across all flight-sim style games to pick what you're looking at as the target, for e.g. joystick players who may not have a hand free to use the mouse. Conversely, what @azieba has implemented appears to be a textbook-perfect version of the "cycle hostiles" feature that many space games (e.g. Elite Dangerous, Star Citizen, etc) also supply to the user. Thus, the PR as written is functionally good to merge (pending code review of course), but just needs to be separated out into its own key so that the user can also have a "select what I'm looking at" button in the future as well. :smile:

@azieba 
> I think I rediscovered that programming can be fun and give satisfaction.

I'm glad to hear it! Game programming is in my experience extremely fun and rewarding, as you get to exercise your creative design skills as well as build on your technical skills.

> Should I switch the PR to draft?

That's perfectly fine, just post here once you open it back up for review!

**Comment by azieba** [comment]
I have implemented "select what I'm looking at" feature. Waiting for feedback.

File: src/Sensors.cpp
Comment (Web-eWorks) [inline src/Sensors.cpp:70]: dynamic_cast is not particularly performant (depending on platform and compiler, it can walk the entire class inheritance tree). It's not a problem when called once per frame, but in game contexts (Pioneer specifically) it's *usually* a sign of bad architecture.

This is not a requirement to merge at all, but I'd recommend that this function be refactored to take in a Body* of the current combat target and return a Body* of the new combat target; the actual assignment to the player's combat target should be performed in the calling function instead.

**Review by Web-eWorks — COMMENTED** [review]
Overall I'm fairly happy with how the code looks, I've pointed out a few things that could potentially be an issue in practice, but I haven't gotten a chance to sit down test this PR yet. Hopefully will get to do that this evening.

**Comment by Web-eWorks** [comment]
After playing with the target selection specifically, it seems this is not without its edge cases. I think probably the most user-intuitive way to solve this would be to hook into some of the data used by the worldview UI to draw object labels.

As a player, if I look at Neptune and hit 'select target' I don't expect to select a random ship 11AU away en route to Neptune that I cannot otherwise see on the screen. Similarly, I'd expect to have selected Neptune rather than Nereid, as while Nereid is certainly closer it's "less important" to the player than Neptune itself.

https://user-images.githubusercontent.com/4218491/204431031-9e55c760-c019-4281-80b3-0278d39d123c.mp4

The code that is responsible for grouping bodies that way is in [`src/lua/LuaPiGui.cpp`](https://github.com/pioneerspacesim/pioneer/blob/8e09b51da787d36103b22a967a9b5e0d556d8357/src/lua/LuaPiGui.cpp#L1945), though making it accessible for code running in the PlayerShipController is a non-trivial refactor and I won't ask you to do that for this PR. I would also like to avoid running that code more than once per frame, as it's duplicated effort and the cost scales with the complexity of the system.

As a stopgap, I would recommend a simple distance filter for selecting ships; the value of `IN_SPACE_INDICATOR_SHIP_MAX_DISTANCE = 1000000` in `pigui/views/game.lua` should be both consistent and sufficient.

Determining which system body to select when multiple are grouped under the cursor is more difficult, but I would recommend selecting the hierarchically-closest body to the system root (walk the parent chain of either the SystemBody or Frame associated with the body) except if the player is directly in the Frame of that body. This would handle edge cases when a moon lines up with the planet itself during orbital maneuvers, but still default to selecting the planet when not operating directly within its frame of influence..

File: src/Space.cpp
Comment (azieba) [reply src/Space.cpp:542]: At this moment it is not called in every game loop, just for the key press so I did not bother much with performance of this function.

**Comment by azieba** [comment]
> As a stopgap, I would recommend a simple distance filter for selecting ships; the value of `IN_SPACE_INDICATOR_SHIP_MAX_DISTANCE = 1000000` from `pigui/views/game.lua` should be both consistent and sufficient.
>

Yup selecting invisible ships might be strange, yet I wanted to be able to select ships that I see on the radar. This is something I would like to have personally.
Maybe the ship select distance should be the same as the max radar distance?

> Determining which system body to select when multiple are grouped under the cursor is more difficult, but I would recommend selecting the hierarchically-closest body to the system root (walk the parent chain of either the SystemBody or Frame associated with the body) except if the player is directly in the Frame of that body. This would handle edge cases when a moon lines up with the planet itself during orbital maneuvers, but still default to selecting the planet when not operating directly within its frame of influence..

I was thinking that maybe it would be better to allow player to decide what he wants to select instead figuring out what he might want? How about implementing target cycling? Like first time you select nearest body then the next further away and so on. This wold be particularly useful when approaching grouped stack of stuff to scoop.

**Comment by Zireael07** [comment]
> Maybe the ship select distance should be the same as the max radar distance?

I'd definitely expect that to be the case.

+1 to target cycling too

**Comment by Web-eWorks** [comment]
> Yup selecting invisible ships might be strange, yet I wanted to be able to select ships that I see on the radar. This is something I would like to have personally.
Maybe the ship select distance should be the same as the max radar distance?

As a compromise, I'd like to expand the radar into its own interface / gameplay loop, allowing players to scan/lock radar contacts and convert them into "target tracks" (which show up as labeled ships in the world view). For now, feel free to set the maximum distance to whatever you like, though if you select something larger than the "visible ships" radius please leave a comment in the code mentioning that for future contribution effort.

> How about implementing target cycling?

Agreed, though I'd still recommend that target cycling should start with the parent body in situations like the above to be intuitive (i.e. should start with Neptune first, then select Nereid, etc.) The Principle of Least Surprise applies heavily in UI / UX design.

File: src/ship/PlayerShipController.cpp
Comment (azieba) [reply src/ship/PlayerShipController.cpp:788]: What about a case when we have a combat target and we select none Ship object, do we deselect the combat target?

Comment (Web-eWorks) [reply src/ship/PlayerShipController.cpp:788]: Because the player is allowed to have two simultaneous selections (combat + nav target) I'd suggest that it simply select the object as the nav target and leave the combat target alone. If needed, we can add another key to swap between treating the combat and nav target as the "active" for whatever context that's needed in.

**Comment by azieba** [comment]
> walk the parent chain of either the SystemBody or Frame associated with the body

Do those hierarchies mirror each other?

**Comment by Web-eWorks** [comment]
> Do those hierarchies mirror each other?

With one caveat, yes. StarSystem and SystemBody are the "offline" representation of Space and Body for systems that are not currently loaded and playing in. Each SystemBody is (generally) converted into a Body which receives its own Frame, though planetary / rotating bodies have *two* Frames, a "non-rotating" frame that contains all child Bodies and a "rotating" frame that contains all bodies within the atmospheric sphere of influence (generally speaking, bodies that would be significantly affected by the rotating surface of the planet).

**Comment by azieba** [comment]
I'm looking for a way to share a constant defined in lua with cpp code. Any example?

**Comment by Web-eWorks** [comment]
Copy and paste :smile:

Unfortunately that's the only way to do it that doesn't get extremely complicated, just copy the value into C++ and leave a note that it's intended to be kept in sync with the Lua variable. The whole "sensors gameplay" aspect of the code is going to be rewritten / upgraded at some point anyways, so if the value needs to be changed at any point that's when we'll change it.

**Comment by azieba** [comment]
@Web-eWorks I think I have implemented all the requested changes. Should I squash the commits?

Regarding sharing of constants between c++ and lua: How about a script that generates header file for c++ and for lua from common list of constants during compilation process? Maybe just lua file from h file or the other way around?

**Review by Web-eWorks — CHANGES_REQUESTED** [review]
Overall I'm extremely happy with the results of this PR - I've left two minor change requests; and there's one bug left to fix: when you're cycling through a target group that contains both nav and combat targets, you can no longer cycle through the nav targets once you've selected the combat target. I've attached a video to demonstrate.

https://user-images.githubusercontent.com/4218491/205455387-55ad4c52-b348-486d-9c27-cf772e7f217e.mp4

As for cleaning up the git history, this is pretty much clean - we try to stick to a "one commit per [thing]" rule, so minor bugfixes get squashed back into the commit they're amending and larger changes stay as their own independent commits. It makes review easier, and if we have to go back in the future and look at the git history it tends to preserve the development process and reasoning better.

**Comment by azieba** [comment]
I can't figure it out. Where and when scan_enums.py is called?

**Comment by Web-eWorks** [comment]
> I can't figure it out. Where and when scan_enums.py is called?

Manually. The primary reason we don't have any kind of complicated build system scripting to autogenerate code at compile time is because the time invested in the tooling soon far outweighs the time it would save the developer doing it manually. There's an [XKCD for this](https://xkcd.com/1319/) that illustrates the problem.

I've suggested you copy-paste the value not because it's impossible to share the value between C++ and Lua, but because it's not worth the time it would take to implement for multiple reasons:
- The entire Sensors code structure is only partially useful and is on the short-list to be thrown out and rewritten when we next focus on sensor-based gameplay (unknown how soon that will be, possibly in the next patch cycle if enough developers are available). A primary focus of that rewrite will be better Lua interop, as well as performance improvements.
- Any codegen based solution (e.g. scan_enums.py) is going to require the game binary to be recompiled to have an updated value anyways. Ctrl+Shift+F takes so little time that you would have to change the same value 100x over to outweigh the developer time cost involved in developing and maintaining an "automated" solution.
- Finally, the actual *proper* way to do this isn't to code-gen any constants or values, but to drive the radar sensors range in both Lua and C++ directly from an equipment property defined in Lua pulled from the ship's equipment loadout. That touches enough files and breaks through enough abstraction layers that I don't want to ask or expect you to try to implement that as a newer developer.
- Especially when it's not worth doing **right now** because there is no supporting gameplay (e.g. different equipment, limited map sensor range, frequency scanning, etc) that would make that effort provide a tangible benefit. Yes, limiting the target selection to radar-visible contacts is "supporting gameplay" but there's nothing else currently implemented that can deal with the idea of a changing maximum radar range.

That rant over, I'd recommend from a number of years of Pioneer development experience to put a pin in the idea of variable sensor ranges, write it down somewhere, and keep it in mind for another pass sometime in the near future as part of a larger gameplay-driven effort.

Not only will you benefit from having increased experience with the Pioneer codebase, but when you write something like this *ex-nihilo*, it usually winds up having to be rewritten when you actually implement the features that would drive the value (e.g. per-equipment radar range, radar contact masking from planets, unidentified / partially identified contacts, etc.)

Hopefully I haven't come across as too draconian or dampened your spirits, just wanted to elucidate on *why* doing it "the definitionally correct way" can actually be worse than doing it the quick way in the long run. I'd rather you maintain your energy and joy of working on Pioneer and work your way into big architectural concerns like that, than dump you straight into the deep end of cutting across 5+ game systems.

**Comment by azieba** [comment]
> Hopefully I haven't come across as too draconian or dampened your spirits

Not really. I suspect this rant is in some part directed at more general dev public. I kind of feel your pain, because I have to work with code "automatically" generated out of excel spreadsheet.
But where do you want to go with this scan_enums.py? It seems that it is still maintained and it was in the build system, but not any more. There are quiet a few constants generated by this script and they are used in lua. If all the constants are to be used maybe it should be brought back into build? It was surprising to me that it is not there.
Personally I was just curious what it was and if I can use it and how. Now that I know my itch is already scratched.
I totally agree that the max radar range should come from equipment definition file and should be available both in C++ and Lua as an equipment property. But how about IN_SPACE_INDICATOR_SHIP_MAX_DISTANCE? Should it also be configurable?

File: src/ship/PlayerShipController.cpp
Comment (azieba) [reply src/ship/PlayerShipController.cpp]: Done

File: src/ship/PlayerShipController.cpp
Comment (azieba) [reply src/ship/PlayerShipController.cpp]: Corrected

**Comment by azieba** [comment]
@Web-eWorks I believe I have fixed the bug and applied the requested corrections. The target selection logic gets complicated with those separate nav and combat targets and it is difficult to decide what the Player would really want. At this point I wanted something to be "moving" after each key press, so e.g. if view is pointed at a single nav body and single ship and there is nothing new to cycle to, then the combat target will be toggled. If we point just to a single nav target, it will be toggled instead.

**Comment by Web-eWorks** [comment]
> There are quite a few constants generated by this script and they are used in lua. If all the constants are to be used maybe it should be brought back into build? It was surprising to me that it is not there.

We run `./scripts/scan_enums.py -o src/enum_table.cpp --pattern='*.h' -r src/` manually only when we've added or changed a C++ enum value that is intended to be exposed to Lua. There's not really any need to run it from CMake, because the outputs are checked in to source control and we have a very low frequency of actually changing the enumerations (they're intended to be "stable" as much as possible, as refactoring Lua to match changes to the enum names is not a simple operation).

> But how about IN_SPACE_INDICATOR_SHIP_MAX_DISTANCE? Should it also be configurable?

This value will (off the top of my head) likely be removed and its effect driven from the radar system - when a radar contact becomes "tracked" (see literature about existing conventional radar systems) it will be eligible for an in-world label on its position. Thus, a static constant is no longer appropriate as it may make sense to impose a limit on the maximum number of "trackfiles" the player can maintain based on their equipment quality or crewmember skills.

...I need to write this down somewhere.

**Comment by azieba** [comment]
> We run `./scripts/scan_enums.py -o src/enum_table.cpp --pattern='*.h' -r src/` manually only when we've added or changed a C++ enum value that is intended to be exposed to Lua.

I know that, but I discovered it by finding ancient makefile that had the target enums: with this command. It is barely discoverable for a new contributor.

**Review by Web-eWorks — APPROVED** [review]
Approved. Many thanks for jumping in on this; here's a beautiful vista from Ganymede for your hard work :smile:

image

**Comment by Web-eWorks** [comment]
> It is barely discoverable for a new contributor.

I agree wholeheartedly. Knowledge like this normally goes on the wiki, though it'd be nicer if there was a script or optional CMake target already available to run the full command - unfortunately it's not been enough of a bother to justify spending the time on recently.

**Pull request merged**
